### PR TITLE
Improve types for `electronic_structure.{core/dos}` 

### PR DESCRIPTION
**Pull request by DanielYang59**
## Summary

- Improve types and docstring cleanups for (follow up of #3873): 
    - `electronic_structure.core`
    - `electronic_structure.dos`
- Nest private `classmethod` (which should not be `classmethod` but `staticmethod`) `get_transformation_matrix` inside `get_moment` method 2843854ea848a5a21898593a26acd8dc44059471, rationale:
    - It's used by `get_moment` alone
    - No unit test added for `get_transformation_matrix` (usually I would expose such internal function as private method if I need to unit test it)

### CODE REVIEW

File: pymatgen/electronic_structure/dos.py [hunk 1]
```diff
@@ -1329,21 +1407,21 @@ def get_site_orbital_dos(self, site: PeriodicSite, orbital: str) -> Dos:  # type
...
-            site: Site in Structure associated with CompleteDos.
+            site (PeriodicSite): Site in Structure associated with LobsterCompleteDos.
 
         Returns:
-            A dict {"e_g": Dos, "t2g": Dos} containing summed e_g and t2g DOS
-            for the site.
+            {"e_g": Dos, "t2g": Dos}: Summed e_g and t2g DOS for the site.
         """
-        warnings.warn("Are the orbitals correctly oriented? Are you sure?")
+        warnings.warn("Are the orbitals correctly oriented? Are you sure?")  # TODO: DanielYang: unconditional warning
```
Comment (DanielYang59) [inline pymatgen/electronic_structure/dos.py:1423]: There is a very strong but unconditional warning here.

Comment (DanielYang59) [reply pymatgen/electronic_structure/dos.py]: @JaGeo Do we want to make this warning less strong? Like `Please make sure the orbitals are correctly oriented.`. Or add a condition for such warning (if possible)?

Comment (DanielYang59) [reply pymatgen/electronic_structure/dos.py]: Thanks for the input! I get it now (check the bond spatial alignment relative to the global axes). Can I change the warning message a bit to make it more descriptive and less strong? Is there any downstream operations that depend on the message?

Perhaps `Please make sure the orbitals/bonds are correctly oriented with respect to global axes`? Or do you want to keep it as is?

**Comment by DanielYang59** [comment]
@janosh Apologize for trying to mess up your weekend, just ping you to get this tracked as well and no rush at all. Thanks!

**Review by janosh — APPROVED** [review]
Good stuff! Thanks

**Comment by DanielYang59** [comment]
Thanks! Have a nice weekend ahead!

**Pull request merged**